### PR TITLE
Call update_median_feeds and check_call_orders when bitasset changes

### PR DESCRIPTION
**Pull request by jmjatlanta**
This is a better solution to Bitshares issue #890. This is based on the (safe) assumption that #868 will roll first or at the same time as #890.

* It is based on the branch of Issue #868, thereby removing conflicts
* The logic in update_bitasset_object_options is clearer.

TODO: Need to improve tests

### CODE REVIEW

File: libraries/chain/asset_evaluator.cpp [hunk 1]
```diff
@@ -410,11 +410,22 @@ void_result asset_update_bitasset_evaluator::do_evaluate(const asset_update_bita
...
+   // If the minimum number of feeds to calculate a median has changed
+   // we need to recalculate the median
    bool should_update_feeds = false;
-   if( op.new_options.minimum_feeds != bdo.options.minimum_feeds )
+   if( op.new_options.minimum_feeds != bdo.options.minimum_feeds)
       should_update_feeds = true;
 
+   // We need to call check_call_orders if the settlement price changes after hardfork 890
+   bool after_hardfork_890 = false;
+   if (db.get_dynamic_global_properties().next_maintenance_time > HARDFORK_CORE_890_TIME) {
```
Comment (abitmore) [inline libraries/chain/asset_evaluator.cpp]: `db.get_dynamic_global_properties()` is called twice in this function. We can use a local variable.

File: libraries/chain/asset_evaluator.cpp [hunk 2]
```diff
@@ -447,8 +458,12 @@ static void update_bitasset_object_options(const asset_update_bitasset_operation
       }
    }
 
+   const auto old_feed_price = bdo.current_feed.settlement_price;
    if( should_update_feeds )
       bdo.update_median_feeds(db.head_block_time());
+
+   if (after_hardfork_890 && old_feed_price != bdo.current_feed.settlement_price)
+      db.check_call_orders(asset_to_update);
```
Comment (abitmore) [inline libraries/chain/asset_evaluator.cpp]: 1. This should only be done in `if( should_update_feeds )`.
2. Although not very sure, I think it's better to call `check_call_orders` out of the `modify` lambda. Perhaps let this static function return a `bool` and call `check_call_orders` accordingly.

File: libraries/chain/asset_evaluator.cpp (hunk 1 above)
Comment (jmjatlanta) [reply libraries/chain/asset_evaluator.cpp]: Both were attempting to get the next maintenance time, so I'm throwing that in a local variable. If we need another value from dynamic_global_properties, we can throw it in a local as well.

File: libraries/chain/asset_evaluator.cpp (hunk 2 above)
Comment (jmjatlanta) [reply libraries/chain/asset_evaluator.cpp]: Done, at the expense of adding some things to the capture list. Please verify my understanding of the logic:
IF (after hardfork 890) AND (feed price changed) THEN (check_call_orders)

File: libraries/chain/asset_evaluator.cpp [hunk 3]
```diff
@@ -450,21 +461,29 @@ static void update_bitasset_object_options(
...
    try
    {
       auto& db_conn = db();
+      const fc::time_point_sec& next_maint_time = db_conn.get_dynamic_global_properties().next_maintenance_time;
       const auto& asset_being_updated = (*asset_to_update);
+      bool price_changed = false;
 
-      db_conn.modify(*bitasset_to_update, [&op, &asset_being_updated, &db_conn](asset_bitasset_data_object& bdo) {
-         update_bitasset_object_options(op, db_conn, bdo, asset_being_updated);
+      db().modify(*bitasset_to_update, [&op, &asset_being_updated, &price_changed, &next_maint_time, &db_conn](asset_bitasset_data_object& bdo) {
```
Comment (abitmore) [inline libraries/chain/asset_evaluator.cpp]: Changed to `db()` here, so perhaps still have conflicts.

I was thinking that we can get `next_maint_time` inside the function.

File: libraries/chain/asset_evaluator.cpp (hunk 2 above)
Comment (abitmore) [reply libraries/chain/asset_evaluator.cpp]: > IF (after hardfork 890) AND (feed price changed) THEN (check_call_orders)

Correct.

File: libraries/chain/asset_evaluator.cpp [hunk 4]
```diff
@@ -407,20 +407,31 @@ void_result asset_update_bitasset_evaluator::do_evaluate(const asset_update_bita
  * @param db the database
  * @param bdo the actual database object
  * @param asset_to_update the asset_object related to this bitasset_data_object
+ * @returns true if the feed price is changed
```
Comment (abitmore) [inline libraries/chain/asset_evaluator.cpp]: Can change to return true only if (feed changed) AND (after_hf_890)

File: libraries/chain/asset_evaluator.cpp (hunk 3 above)
Comment (jmjatlanta) [reply libraries/chain/asset_evaluator.cpp]: Fixed

File: libraries/chain/asset_evaluator.cpp (hunk 4 above)
Comment (jmjatlanta) [reply libraries/chain/asset_evaluator.cpp]: Done.

File: libraries/chain/asset_evaluator.cpp [hunk 5]
```diff
@@ -450,8 +462,11 @@ static void update_bitasset_object_options(
       }
    }
 
+   const auto old_feed_price = bdo.current_feed.settlement_price;
    if( should_update_feeds )
       bdo.update_median_feeds(db.head_block_time());
+
+   return next_maint_time > HARDFORK_CORE_890_TIME && old_feed_price != bdo.current_feed.settlement_price;
```
Comment (abitmore) [inline libraries/chain/asset_evaluator.cpp]: I'd code like this:
```
   if( !should_update_feeds )
      return false;

   const auto old_feed_price = bdo.current_feed.settlement_price;
   bdo.update_median_feeds(db.head_block_time());

   return next_maint_time > HARDFORK_CORE_890_TIME && old_feed_price != bdo.current_feed.settlement_price;
```

Comment (jmjatlanta) [reply libraries/chain/asset_evaluator.cpp]: Modified. Tell me if the new way shows intent clearly. Thanks.

**Comment by jmjatlanta** [comment]
This PR made it difficult to work together with other bitshares branches. It was eventually moved to the bitshares repository, where modifications continued.

**Pull request closed without merging**